contracts/javascore/modules/ics20app/src/main/java/ibc/ics20app/ICS20Transfer.java

```diff
@@ -0,0 +1,214 @@
+package ibc.ics20app;
+
+import ibc.icon.interfaces.IIBCModule;
+import ibc.icon.score.util.StringUtil;
+import ibc.ics23.commitment.Ops;
+
+import icon.proto.core.channel.Channel;
+import icon.proto.core.channel.Packet;
+import score.Address;
+import score.Context;
+import score.DictDB;
+import score.annotation.External;
+
+import java.math.BigInteger;
+import java.util.Arrays;
+
+import static ibc.ics20app.ICS20TransferBank.bank;
+
+public abstract class ICS20Transfer implements IIBCModule {
+    public static final String ICS20_VERSION = "ics20-1";
+    public static final Address ZERO_ADDRESS = Address.fromString("hx0000000000000000000000000000000000000000");
+    public static final DictDB<String, Address> channelEscrowAddresses = Context.newDictDB("channelEscrowAddresses", Address.class);
+    protected final DictDB<String, String> destinationPort = Context.newDictDB("destinationPort", String.class);
+    protected final DictDB<String, String> destinationChannel = Context.newDictDB("destinationChannel", String.class);
+
+    @External(readonly = true)
+    public Address getIBCAddress() {
+        return ICS20TransferBank.ibcHandler.getOrDefault(ZERO_ADDRESS);
+    }
+
+    public void onlyIBC() {
+        Context.require(Context.getCaller().equals(getIBCAddress()), "ICS20App: Caller is not IBC Contract");
+    }
+
+    @External(readonly = true)
+    public String getDestinationPort(String channelId) {
+        return destinationPort.get(channelId);
+    }
+
+    @External(readonly = true)
+    public String getDestinationChannel(String channelId) {
+        return destinationChannel.get(channelId);
+    }
+
+    @External
+    public byte[] onRecvPacket(byte[] packet, Address relayer) {
```

Comment:
Since this fucntion is not allowed to fail, perhaps create a internal function and wrap this in a try catch to always return failed ack in case of failure